libbpf-tools/oomkill.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2022 Jingxiang Zeng
+//
+// Based on oomkill(8) from BCC by Brendan Gregg.
+// 13-Jan-2022   Jingxiang Zeng   Created this.
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include "oomkill.skel.h"
+#include "oomkill.h"
+#include "trace_helpers.h"
+
+const char *argp_program_version = "oomkill 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz) {
+	FILE *f;
+	char buf[256];
+	int n;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	struct data_t *e = data;
+
+	f = fopen("/proc/loadavg", "r");
+	if (f) {
+		memset(buf, 0 , sizeof(buf));
+		n = fread(buf, 1, sizeof(buf), f);
+		fclose(f);
+	}
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	
+	if (n)
+		printf("%s Triggered by PID %d (\"%s\"), OOM kill of PID %d (\"%s\"), %lld pages, loadavg: %s\n",
+			ts, e->fpid, e->fcomm, e->tpid, e->tcomm, e->pages, buf);
+	else
+		printf("%s Triggered by PID %d (\"%s\"), OOM kill of PID %d (\"%s\"), %lld pages\n",
+                        ts, e->fpid, e->fcomm, e->tpid, e->tcomm, e->pages);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	printf("Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+int main(int argc, char **argv) {
+	struct perf_buffer *pb = NULL;
+	struct oomkill_bpf *obj;
+	int err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	libbpf_set_print(libbpf_print_fn);
+
+	obj = oomkill_bpf__open_and_load();
+	if (!obj) {
+		fprintf(stderr, "failed to load and open BPF object\n");
+		return 1;
+	}
+
+	err = oomkill_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), 64,
+			handle_event, handle_lost_events, NULL, NULL);
+	if (!pb) {
+		err = -errno;
+		fprintf(stderr, "failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	printf("Tracing OOM kills... Ctrl-C to stop.\n");
+
+	while ((err = perf_buffer__poll(pb, 100)) >= 0)
```

Comment:
Please install a signal handler and handle error in this loop. Refer to
https://github.com/iovisor/bcc/blob/master/libbpf-tools/mountsnoop.c#L291-L311

Reply:
Fix it in new commit 89ce26bd6ca2c8a689042f76ea3e4340d61536a8